crates/sled/src/tree.rs

```diff
@@ -186,22 +186,22 @@ impl Tree {
     /// let t = sled::Tree::start(config).unwrap();
     ///
     /// // unique creation
-    /// assert_eq!(t.cas(vec![1], None, Some(vec![1])), Ok(()));
-    /// // assert_eq!(t.cas(vec![1], None, Some(vec![1])), Err(Error::CasFailed(Some(vec![1]))));
+    /// assert_eq!(t.cas(&[1], None as Option<&[u8]>, Some(vec![1])), Ok(()));
```

Comment:
This is a breaking change, the `cas` method now accept an `Option<AsRef<[u8]>>` and this can be a problem for the inference system. If you call `cas` with a `None` in place of the `old` argument you must specify which type is the `Option`.

A solution could be to continue accepting an `Option<&[u8]>`.

Reply:
Grrr https://github.com/rust-lang/rust/issues/20063 strikes again... In this case, I'd prefer to keep the signature for cas the way it was. The old value is flexible when it stays as an `Option<&[u8]>`, and we need to allocate anyway to fill the new value if it's `Some`. The clone that we do while creating the node can be optimized away by using an `Option` above the loop that is set before the loop with the provided value, taken to fill the new node, and if the inner CAS fails we can take the value from the created node and reuse it in the next loop.

No need to implement the zero-additional-copy `Option` trick, but I think it's better to keep the API as it is for `cas` because it communicates the need to allocate. The other changed methods need to allocate too, and they should also implement this trick to remove their inner `clone` calls, but I'm OK with them accepting any `AsRef<[u8]>` because it makes trying out the system much easier. 

It's a bit unidiomatic to not require an owned value to be passed in, but in this case it does let us simplify usage in some cases, especially for new people who are just trying it out, and I want to make this easier where possible. It's at conflict with the `cas` method staying explicit and owned, but I'm OK with that in this case.

Reply:
I just though about `Into<Vec<u8>>` combined with `AsRef<[u8]>`, but it will complexify the API a little bit. Using these bounds we could avoid duplications if the user gives a `Vec<[u8]>` or a `Box<[u8]>` as key but will allocate if it's a raw `&[u8]` for example.

In this case doesn't a combination of `Borrow` and `Eq` better fitting ?
Like the [`HashMap::get`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get) method.

I think we should just make this PR effective now and think about that (or not) in another place.